### PR TITLE
Adds single node allocation to shrink

### PR DESCRIPTION
**Pull request by colings86**
This change adds two new steps as the first steps of the shrink action:

1. A `SetSingleNodeAllocateStep` which:
    1. Determines which of the active nodes match the existing index
allocation rules
    2. Randomly (using Randomness so its deterministic for testing)
picks one of the matching nodes
    3. Updates the index settings to add a require allocation rule for
the node that was picked (using the
`index.routing.allocation.require._name` setting)
2. An `AllocationRoutedStep` which ensures that at least one copy of
each shard is allocated according to the new allocation rules

Note that this change also modifies the `AllocationRoutedStep` to add a
boolean field which determines whether the allocation is complete when
at least one copy of each shard matches the allocation rulees or if it
needs to wait for all shard copies to be allocated according to the
rules.

Lastly, a `randomStepKey()` method is added to `AbstractStepTestCase`
for convenience.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/AsyncActionStep.java [hunk 1]
```diff
@@ -25,7 +26,7 @@ public boolean indexSurvives() {
         return true;
     }
 
-    public abstract void performAction(IndexMetaData indexMetaData, Listener listener);
+    public abstract void performAction(IndexMetaData indexMetaData, ClusterState currentClusterState, Listener listener);
```
Comment (talevy) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/AsyncActionStep.java]: should we just have `Index` and `ClusterState` and fetch `indexMetaData` internally every time?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/SetSingleNodeAllocateStep.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
+/*
```
Comment (talevy) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/SetSingleNodeAllocateStep.java]: watch license

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/indexlifecycle/SetSingleNodeAllocateStepTests.java [hunk 3]
```diff
@@ -0,0 +1,425 @@
+/*
```
Comment (talevy) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/indexlifecycle/SetSingleNodeAllocateStepTests.java]: watch license

**Review by talevy — CHANGES_REQUESTED** [review]
should we add a test to IndexLifecycleRunnerTests with a real clusterState object, and verify that the indexMetaData that was passed in, is in fact, sameAs the one found inside the clusterState argument?

**Comment by colings86** [comment]
@talevy I pushed a commit to address your comments. Regarding the test you suggested, I don't think IndexLifecycleRunnerTests would be the right place for that since the IndexMetadata is retrieved from the  ClusterState before that point (in IndexLifecycleService). I looked into adding it to IndexLifecycleServiceTests but I'm not sure that is the right place either.

**Review by talevy — APPROVED** [review]
LGTM.

Then that is OK, I thought there was a place for testing those parameters.

**Pull request merged**